<?xml version="1.0" encoding="utf-8"?>

<launch>
  <!-- ======== USB CAMERA Node ======== -->
  <node pkg="usb_cam" type="usb_cam_node" name="usb_cam" output="screen">
    <param name="video_device" value="/dev/video2"/>
    <param name="image_width" value="640"/>
    <param name="image_height" value="480"/>
    <param name="pixel_format" value="yuyv"/>
    <param name="io_method" value="mmap"/>
    <param name="camera_frame_id" value="usb_cam"/>
  </node>

  <!-- ======== YOLOv3 Object Detection (darknet_ros) ======== -->
  <arg name="network_param_file" default="$(find darknet_ros)/config/yolov3.yaml"/>
  <include file="$(find darknet_ros)/launch/darknet_ros.launch">
    <arg name="network_param_file" value="$(arg network_param_file)"/>
    <arg name="image" value="/usb_cam/image_raw"/>
  </include>

  <!-- ======== Face Detection and Recognition ======== -->
  <arg name="haarcascade" default="$(find face_detection_recognition)/data/haarcascade_frontalface_default.xml"/>
  <arg name="encodings" default="$(find face_detection_recognition)/data/encodings.pickle"/>

  <node pkg="face_detection_recognition"
        type="face_recognizer.py"
        name="face_recognizer"
        output="screen">
    <param name="haar_path" value="$(arg haarcascade)"/>
    <param name="input_image" value="/usb_cam/image_raw"/>
    <param name="output_image" value="/face_recognizer/output_image"/>
    <param name="tolerance" value="0.6"/>
  </node>

  <node pkg="rqt_image_view"
        type="rqt_image_view"
        name="face_recognizer_view"
        args="/face_recognizer/output_image"
        output="screen"/>

  <!-- ======== Emotion and Gender Detection ======== -->
  <node name="emotion_gender_detector_node"
        pkg="emotion_gender_detector"
        type="perception_node.py"
        output="screen">
    <param name="image_topic" value="/usb_cam/image_raw"/>
    <param name="face_detector_backend" value="opencv"/>
  </node>

  <node name="rqt_debug_image_view"
        pkg="rqt_image_view"
        type="rqt_image_view"
        args="/debug_image"/>

  <!-- ======== Unified Detection Logger Node ======== -->
  <node name="detection_logger_node"
        pkg="face_detection_recognition"
        type="detection_logger_node.py"
        output="screen"/>
</launch>
